### PR TITLE
chore: use pydantic for yaml test validation

### PR DESCRIPTION
**Pull request by terrykong**
We've had nightly tests fail because the yaml configs don't adhere to the appropriate `MasterConfig`. This introduces `pydantic` in the unit tests to validate all our configs.

Introducing has helped fill in some missing config fields.

We don't use pydantic to validate at runtime since that may be jarring to use an old config with extra keys since we do not want to forbid users from using a config structure that works for them. Instead, we only use this for testing to make sure the current configs in the repo adhere to the current state of the code. This way, if a user changes a config value from `NotRequired[int]` to `int`, we should be able to catch that now on the entire `MasterConfig`.

This also includes some small config fixes:
1. removing stop_token_ids from the configs since if someone swaps out the model, it may be surprising that the stop tokens are kept from the previous tokenizer
2. add a check for defer_fp32_logits==True if logprob_chunk_size>0

## Summary by CodeRabbit

* **New Features**
  * Added training optimization configuration options including bias activation fusion, optimizer offloading, and custom FSDP settings.
  * Expanded configuration flexibility to allow optional values for resource allocation and generation parameters.

* **Bug Fixes**
  * Updated padding token handling in generation workflows.
  * Fixed default values for optimization flags.

* **Configuration Updates**
  * Removed explicit stop token IDs from generation configs for improved default termination behavior.
  * Enhanced type safety for optional configuration parameters.

### CODE REVIEW

File: tests/unit/test_config_validation.py [hunk 1]
```diff
@@ -14,225 +14,123 @@
...
-        if key not in all_valid_keys:
-            errors.append(f"Extra key in {section_path}: {key}")
-
-    return errors
 
+        config_info = f"\n\nConfig file: {config_file}" if config_file else ""
+        raise AssertionError(
+            f"Config validation failed for section '{section_name}':{config_info}\n"
+            + "\n".join(error_messages)
+        )
```
[the remark this answers is not in the document]
Comment (terrykong) [reply tests/unit/test_config_validation.py:83]: good point. fixed 50d0d87c39e62cd5dd2dfbec926e788e618b8bca

File: tests/unit/test_recipes_and_test_suites.py [hunk 2]
```diff
@@ -36,6 +36,7 @@
     "grpo": "examples/configs/grpo_math_1B.yaml",
     "vlm_grpo": "examples/configs/vlm_grpo_3B.yaml",
     "distillation": "examples/configs/distillation_math.yaml",
+    "rm": "examples/configs/rm.yaml",
```
[the remark this answers is not in the document]
Comment (terrykong) [reply tests/unit/test_recipes_and_test_suites.py:39]: wrong. it's used on line 243

File: tests/unit/test_config_validation.py [hunk 3]
```diff
@@ -14,225 +14,123 @@
...
 
-        # If the annotation is a TypedDict (nested config), validate it recursively
-        if hasattr(annotation, "__annotations__") and isinstance(value, dict):
-            # This is a nested TypedDict, validate it recursively
-            nested_errors = validate_nested_config_section(
-                value, annotation, current_path
+def validate_config_section(
+    section_config: Dict[str, Any],
+    config_class: Type,
+    section_name: str,
```
Comment (ashors1) [inline tests/unit/test_config_validation.py]: I'm slightly unclear on the purpose of `section_name` here because it seems to imply that we're only validating one section of the config (e.g. only validating the `grpo` section for grpo configs). Is it only used to print a more informative error? If so, maybe we could rename it to something like `algorithm_name` for clarity?

Comment (terrykong) [reply tests/unit/test_config_validation.py]: ah, you're right. sorry i accidentally kept that from the prev iteration. it's not needed anymore:

061d4da630902d3fe3187e661d075f597194449f

File: nemo_rl/data/__init__.py [hunk 4]
```diff
@@ -40,6 +43,110 @@ class DataConfig(TypedDict):
...
+        "mmlu_ID-ID",
+        "mmlu_IT-IT",
+        "mmlu_JA-JP",
+        "mmlu_KO-KR",
+        "mmlu_PT-BR",
+        "mmlu_ZH-CN",
+        "mmlu_SW-KE",
+        "mmlu_YO-NG",
+    ]
+    shuffle: NotRequired[bool]
```
Comment (yuki-97) [inline nemo_rl/data/__init__.py]: `shuffle` is always set to False for eval datasets, so I think we can remove `shuffle` in all eval dataset configs.

others LGTM.

Comment (terrykong) [reply nemo_rl/data/__init__.py]: f74640a4fed5c886bbcbd8745fd42a2e495c9044
